go/libraries/doltcore/doltdb/commit_hooks.go

```diff
@@ -93,6 +94,97 @@ func pushDataset(ctx context.Context, destDB, srcDB datas.Database, tempTableDir
 	return err
 }
 
+type PushArg struct {
+	ds   datas.Dataset
+	db   datas.Database
+	hash hash.Hash
+}
+
+type AsyncPushOnWriteHook struct {
+	out io.Writer
+	ch  chan PushArg
+}
+
+const asyncPushBufferSize = 100
+
+var _ datas.CommitHook = (*AsyncPushOnWriteHook)(nil)
+
+// NewAsyncPushOnWriteHook creates a AsyncReplicateHook
+func NewAsyncPushOnWriteHook(ctx context.Context, destDB *DoltDB, tmpDir string) *AsyncPushOnWriteHook {
+	ch := make(chan PushArg, asyncPushBufferSize)
+
+	var newHeads = make(map[string]PushArg, asyncPushBufferSize)
+	go func() error {
+		defer close(ch)
+		for {
+			p, ok := <-ch
+			if !ok {
+				return ctx.Err()
+			}
+			newHeads[p.ds.ID()] = p
+		}
+	}()
+
+	go func() error {
```

Comment:
Not sure when this routine is gonna exit.  It is dependent on ctx.Done to exit.  Is that what we want?  What context is this using?